### ISSUE TITLE
Allow MemberExpressions to bump indentation body by one indentation mark

### ISSUE THREAD

**Issue by mlewand** [eslint/eslint#8426]
**What rule do you want to change?**

indent

**Does this change cause the rule to produce more or fewer warnings?**

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

More permisive logic.

**Please provide some example code that this change will affect:**

Currently when I'm using `"indent": [ "error", "tab" ]` rule, it will complain about following code:

```js
let x = new Promise( ( resolve, reject ) => {
		if ( window.abc ) {// ESLint expects one tab starting from here.
			resolve()
		} else {
			reject();
		}
	} )// all the way until thils line.
	.then(() => {
		return 2;
	} )
	.then(() => {
		return 1;
	} );
```

So the request here is to allow indentation to be bumped by one when `MemberExpression` is open.

This style makes more readable code in case when assigning to a variable, returning, throwing etc.

**What does the rule currently do for this code?**

With current ESLint logic it expects the following:

```js
let x = new Promise( ( resolve, reject ) => {
	if ( window.abc ) {
		resolve()
	} else {
		reject();
	}
} )
	.then(() => {
		return 2;
	} )
	.then(() => {
		return 1;
	} );
```

**What will the rule do after it's changed?**

Allowing following code:

```js
let x = new Promise( ( resolve, reject ) => {
		if ( window.abc ) {
			resolve()
		} else {
			reject();
		}
	} )
	.then(() => {
		return 2;
	} )
	.then(() => {
		return 1;
	} );
```

**Comment by not-an-aardvark** [comment]
Thanks for the report. However, this is working as intended -- the rule expects the opening `{` in your `(resolve, reject)` function to have the same indentation as your closing `{`.

**Comment by mlewand** [comment]
Thanks for reply @not-an-aardvark so you're saying that I should use additional rule to enforce opening  `{` to sit in the same line as arguments?

Because you have to admit, this doesn't look any good:

```
( function() {
	return new Promise( ( resolve, reject ) => {
		if ( window.abc ) {
			resolve()
		} else {
			reject();
		}
	} )
		.then(() => {
			return 2;
		} )
		.then(() => {
			return 1;
		} );
} )();
```

**Comment by not-an-aardvark** [comment]
Another option would be to format the code like this:

```js
( function() {
	return new Promise( ( resolve, reject ) => {
		if ( window.abc ) {
			resolve()
		} else {
			reject();
		}
	} ).then(() => {
		return 2;
	} ).then(() => {
		return 1;
	} );
} )();
```

**Comment by mlewand** [comment]
I see. I checked all the related options and none of them would allow me to achieve my expected result. The thing is that I find it to be much, much more clear when your `return/let/var/const/yield/throw` statements bump indentation by 1 tab.

Having that said I know that there are pretty much as many (if not more) indentation ideas, as there are programmers in the world 🙂

Although I believe this was supported back in JSCS. Does it make sense for a feature request?

**Comment by not-an-aardvark** [comment]
Sure, feel free to fill out the [rule change proposal template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md) (although I can't guarantee the team will accept the proposal).

Are you sure it's the `return` that makes the code more clear? For example, would your code look different if the statement was like this instead?

```js
( function() {
	foo = new Promise( ( resolve, reject ) => { // this line is no longer a return
			if ( window.abc ) {
				resolve()
			} else {
				reject();
			}
		} )
		.then(() => {
			return 2;
		} )
		.then(() => {
			return 1;
		} );
} )();
```

To me, it seems like the style that you're looking for applies to `MemberExpression`s (i.e. property accesses like `.then` where the part before the `.` is multiline.

**Comment by mlewand** [comment]
Ohh, actually you're completely right. Should I post it simply here, or make a new issue?

**Comment by not-an-aardvark** [comment]
I think editing the original issue would probably be the best thing to do.

**Comment by mlewand** [comment]
@not-an-aardvark done, thank you for the assistance! 🍺

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
